### DISCUSSION TITLE
Port of μP to JAX (Haiku specifically)

### DISCUSSION THREAD

**Discussion by davisyoshida** [jax-ml/jax#10528] [Show and tell]
[μP](https://github.com/microsoft/mup) is a method for reparameterizing NNs so that hyperparameter optima stay fixed as you scale them up. I wrote a [port](https://github.com/davisyoshida/haiku-mup) to JAX here.

As an example, I trained a bunch of transformers on PTB, with and without μP. As you can see, μP causes the optimal learning rate to stay fixed as you scale the models up: 

![image](https://user-images.githubusercontent.com/1377776/166327911-7423b255-f426-4c59-9571-b09e863386de.png)

(In the paper, the authors do this experiment but for LMs with 6.7B parameters).

This is specifically for Haiku, but if people are interesting in making versions for FLAX etc., it might be helpful to take a look at this. I had to change the design substantially from the authors' original repo, since a lot of things which are easy in Torch (via mutability) aren't in JAX, and vice versa. I'm also interest in any suggestions on improving the design or increasing usability.

**Comment by YouJiacheng** [comment]
Great Job!
But Why not implement muP in optax?

**Reply by davisyoshida** [reply]
Well, there is an optax transform in there, but you need to mess with the initialization of variables and add some transformations to the output layer, so it can't be done in optax only.

**Reply by YouJiacheng** [reply]
Yes. I mean that you can implement it in optax, and provide a initializer(per array or initialize the whole pytree at once.)

**Reply by YouJiacheng** [reply]
Besides, you can make the first update of muP optimizer correctly initialize the parameters, though it is a bit tricky.

**Reply by davisyoshida** [reply]
Unfortunately you can't figure out what transformations you need to do from only the tensor shape. You do need to know a little bit about how it's being used to know what dimension is fan-in/fan-out. You could maybe guess from the parameter names, but I preferred to create a custom haiku parameter creator so you can directly check which module is requesting the parameter.

**Reply by YouJiacheng** [reply]
Yes, but that will substantially couple muP library with a NN library.
I preferr to ask user to provide fan-in axis information by a pytree.
And muP library author/community can implement a automatic fan-in axis assigner for each popular NN libraries.
This make muP library portable and extensible.

**Reply by davisyoshida** [reply]
You're right, that's a good idea. I'll take a shot at decoupling them.
